File: app/src/main/java/li/garteroboter/pren/qrcode/QRCodeImageAnalyzer.java
package li.garteroboter.pren.qrcode;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

import com.google.zxing.BinaryBitmap;
import com.google.zxing.ChecksumException;
import com.google.zxing.FormatException;
import com.google.zxing.NotFoundException;
import com.google.zxing.PlanarYUVLuminanceSource;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import com.google.zxing.multi.qrcode.QRCodeMultiReader;

import java.nio.ByteBuffer;

import static android.graphics.ImageFormat.YUV_420_888;
import static android.graphics.ImageFormat.YUV_422_888;
import static android.graphics.ImageFormat.YUV_444_888;




public class QRCodeImageAnalyzer implements ImageAnalysis.Analyzer {

    private QRCodeFoundListener listener;

    public QRCodeImageAnalyzer(QRCodeFoundListener listener) {
        this.listener = listener;
    }

    @Override
    public void analyze(@NonNull ImageProxy image) {
        // Log.i(TAG, "Camera Log : Analyze");

        if (image.getFormat() == YUV_420_888 || image.getFormat() == YUV_422_888 || image.getFormat() == YUV_444_888) {
            ByteBuffer byteBuffer = image.getPlanes()[0].getBuffer();
            byte[] imageData = new byte[byteBuffer.capacity()];
            byteBuffer.get(imageData);

            PlanarYUVLuminanceSource source = new PlanarYUVLuminanceSource(
                    imageData,
                    image.getWidth(), image.getHeight(),
                    0, 0,
                    image.getWidth(), image.getHeight(),
                    false
            );

            // could maybe optimize, reduce size and grayscale
            // Log.i(TAG, String.format("PlanarYUVLuminanceSource width = %d height =  %d " , source.getWidth(), source.getHeight()));

            BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(source));


            try {
                Result result = new QRCodeMultiReader().decode(binaryBitmap);
                listener.onQRCodeFound(result.getText());
            } catch (FormatException | ChecksumException | NotFoundException e) {
                listener.qrCodeNotFound();
            }
        }

        image.close();
    }
}

File: app/src/main/java/li/garteroboter/pren/qrcode/qrcode/QrCodeImageAnalyzer.kt
package li.garteroboter.pren.qrcode.qrcode

import android.graphics.ImageFormat
import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.google.zxing.BinaryBitmap
import com.google.zxing.PlanarYUVLuminanceSource
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.qrcode.QRCodeReader

class QRCodeImageAnalyzer(private val listener: QRCodeFoundListener? = null) : ImageAnalysis.Analyzer {

    private val yuvFormats = listOf(ImageFormat.YUV_420_888, ImageFormat.YUV_422_888, ImageFormat.YUV_444_888)


    override fun analyze(image: ImageProxy) {

        if (image.format in yuvFormats) {

            val byteBuffer = image.planes[0].buffer
            val imageData = ByteArray(byteBuffer.capacity())
            byteBuffer[imageData]
            val source = PlanarYUVLuminanceSource(
                imageData,
                image.width, image.height,
                0, 0,
                image.width, image.height,
                false
            )

            val binaryBitmap = BinaryBitmap(HybridBinarizer(source))
            try {
                val result = QRCodeReader().decode(binaryBitmap)
                listener?.onQRCodeFound(result.text)
            } catch (e: Exception ) {
                listener?.qrCodeNotFound()
            }
        } else {
            Log.e(TAG, "FATAL: image.format not in yuvFormats")
        }
        image.close()
    }

    companion object {
        const val TAG = "QRCodeImageAnalyzer"
    }

}